src/arch/amdgpu/vega/pagetable_walker.cc

```diff
@@ -101,6 +101,45 @@ Walker::WalkerState::startFunctional(Addr base, Addr vaddr,
 }
 
 
+void Walker::insert(Addr paddr, PageTableEntry entry)
+{
+    paddr_pte_t *newEntry = nullptr;
+    if (!freeList.empty()) {
+        newEntry = freeList.front();
+        freeList.pop_front();
+    } else {
+        newEntry = entryList.back();
+        entryList.pop_back();
+    }
+    DPRINTF(GPUPTWalker, "Insert pte for paddr %#x. Current length: %d\n",
+            paddr, entryList.size());
+    *newEntry = std::make_pair(paddr, entry);
+    entryList.push_front(newEntry);
+}
+PageTableEntry* Walker::lookup(Addr paddr)
+{
+    auto entry = entryList.begin();
+
+    for ( ; entry != entryList.end(); ++entry)
```

Comment:
please add a comment explaining what this is doing

Reply:
```suggestion
    for (auto entry : entryList) {
```
Something like that would be better and more standard modern C++

Reply:
@TomXia : still need to make change @powerjg requested here

Reply:
Thanks for pointing out.
To track entries in LRU order, the container might be modified. Range based for might not be a direct way to implement it. Shall I still proceed to change?

Reply:
How does the change impact LRU order?

Reply:
entryList (now renamed to pwcEntryList) here maintains the LRU-ordered list of current cached items. Thus, the list needs to be updated in this lookup method. Using regular for loop, it is simper to reorder the list within the loop because the iterators automatically come with it.

Although, there might be a different and efficient way to achieve it with range-based for that I am not aware of.

Reply:
Maybe I'm just missing something, but I'm asking for why this:

for ( ; entry != entryList.end(); ++entry)

and this:

for (auto entry : entryList) {

would lead to a different update order.  To me, they seem like they would both result in the same order ... I'm pretty sure the second option is the same as the first but using newer C++ primitives?

Reply:
Here's the problem with this code: The comment isn't correct. It's not "searching the buffer for the paddr." It's updating the LRU position of an entry. If the code did what the comment said, then it would be read-only and a range-based for loop would work. What you're actually doing is manipulating the list inside an iteration (which is error prone, but looks OK in this case). As you said @TomXia you probably can't use the range-based for given you're changing the list.

(I will also go back to my first comment which is if you used the `AssociativeCache` from base/ you wouldn't have to worry about any of this.)